representation/instance.metta

```diff
@@ -12,29 +13,72 @@
 (: InstanceSet (-> $score Type))
 (: mkSInstSet (-> (List (ScoredInstance $score)) (InstanceSet $score)))
 
-;; sortDeme -- sort scored instances in a deme by the composite score in a decreasing order
-;; this is a wrapper for the actual sorting function sortSInst
+;; --- Conversion helpers ---
+(: instToExpr (-> Instance Expression))
+(= (instToExpr (mkInst $lst)) (List.listToExpr $lst))
+
+(: exprToInst (-> Expression Instance))
+(= (exprToInst $expr) (mkInst (exprToList $expr)))
+
+(: sInstToExpr (-> (ScoredInstance $score) Expression))
+(= (sInstToExpr (mkSInst (mkPair $inst $score)))
+    (let* (
+      ($instExpr (instToExpr $inst))
+      ($scoreExpr (cScoreToExpr $score))
+    )
+      (cons-atom $instExpr $scoreExpr)))
+
+(= (exprToSInst $expr)
+    (let* (
+      (($instExpr $scoreExpr) (decons-atom $expr))
+      ($inst (exprToInst $instExpr))
+      ($score (exprToCScore $scoreExpr))
+    )
+      (mkSInst (mkPair $inst $score))))
+
+(= (sInstSetToExpr (mkSInstSet $lst)) (sInstSetListToExpr $lst))
+
+(= (sInstSetListToExpr $lst)
+    (List.listToExpr (List.map sInstToExpr $lst)))
+
+(: exprToSInstSet (-> Expression (InstanceSet Cscore)))
+(= (exprToSInstSet $expr)
+    (mkSInstSet (List.map exprToSInst (exprToList $expr))))
+
+;; --- Fast selectionSort implementation with conversion ---
 (: sortDeme (-> (InstanceSet Cscore) (InstanceSet Cscore)))
-(= (sortDeme (mkSInstSet Nil)) (mkSInstSet Nil))
-(= (sortDeme (mkSInstSet (Cons $scoredInst $rest)))
-    (mkSInstSet (sortSInst (Cons $scoredInst $rest))))
+(= (sortDeme $set)
+    (let* (
+      ($expr (sInstSetToExpr $set))
+      ($n (len $expr))
+      ($sortedExpr (selectionSort $expr $n sInstComparator))
+    )
+      (exprToSInstSet $sortedExpr)))
 
-;; sortSInst -- sorts list of scored instances using insertion sort     
 (: sortSInst (-> (List (ScoredInstance Cscore)) (List (ScoredInstance Cscore))))
```

Comment:
I don't know if you noticed but the `sortSInst`, the `insertSInst` were origianlly helper functions for the `sortDeme`. If the `sortDeme` test cases pass as they were, we won't need them, therefore, you can safely remove those functions along with their test cases.